### PR TITLE
Deployment of cloud endpoints

### PR DESCRIPTION
**Pull request by gaplyk**
This process is not ready yet.
Right now feedbacks are valuable.

The reason: hard to manage endpoints if you have 2 or more environments. And before release, this manual step requires one more commit to update version of endpoints.

This changes is to automate deployment of swagger scheme at propagate returned version to .kube.yml file to `esp` container.
Also checks for previous deployed versions of swagger ( update `version` field from scheme to deploy the new version).

### CODE REVIEW

**Comment by gaplyk** [comment]
One open question is how to track changes to a schema.

**Comment by jrynyt** [comment]
Can you give a summary of what this is? I have a feeling it should be a separate plugin, since Endpoints is not directly related to GKE.

**Comment by gaplyk** [comment]
endpoints relates to .kube.yml file. Maybe it's make sense to make it as a separate plugin and just update overwrite .kube.yml file in pipeline.

**Comment by jrynyt** [comment]
> Can you give a summary of what this is?

**Comment by jrynyt** [comment]
Endpoints does not relate to kubernetes configuration -- it just happens that your deployment contains an Endpoints container whose configuration is in .kube.yml.

You might look at ConfigMaps: https://kubernetes.io/docs/tasks/configure-pod-container/configmap/. It won't get you all the way there until this is done, though: https://github.com/kubernetes/kubernetes/issues/41711. But you could initialize `kubectl` in a build step and set the active Endpoints service version in the config map there.

File: main.go [hunk 1]
```diff
@@ -168,6 +174,62 @@ func wrapMain() error {
...
+			}
+			err = yaml.NewYAMLOrJSONDecoder(f, int(fi.Size())).Decode(&version)
+			if err != nil {
+				return fmt.Errorf("Error: %s\n", err)
+			}
+			f.Close()
+		}
+		data["endpoints_service_name"] = vargs.EndpointsService
+
+		if len(serviceVersion) == 0 || string(swaggerVersion) != version.Info.Version {
```
Comment (tonglil) [inline main.go]: Is this where you are comparing the deployed version with the version specified in the file using the semver of the openAPI spec?

Comment (gaplyk) [reply main.go]: yes

**Comment by tonglil** [comment]
One way to separate this is to extract the code changes you wrote into a standalone binary, which would be used like:

```yaml
pipeline:
  cloud-endpoints:
    - # authenticate to gke cluster
    - deploy-api --spec openapi.yml --configmap configmap.yml # applies a new configmap w/ the values
  drone-gke:
    ... # run esp with values referenced the new configmap
```

Note: the name of the configmap is appended with the git-sha.

The implication is that there will be as many CMs as there are deployments because every deployment creates a new CM (ties a CM to a Deployment, see https://github.com/kubernetes/kubernetes/issues/22368).

**Comment by tonglil** [comment]
Here's another approach using an echo-server example that I think mitigates the issues above, specifically:
- In-place update of a live ConfigMap referenced multiple Deployments, which could cause issues if the new Deployment with the new endpoint fails while old Deployments spin up new Pods.
- Creating a ConfigMap for each deployment, polluting objects and in addition hashes can be cumbersome.

The approach:

In `.kube.yml`:

```yml
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: echo-app
  namespace: esp
spec:
  replicas: 1
  template:
    metadata:
      labels:
        app: echo-app
    spec:
      containers:
      - image: gcr.io/google_containers/echoserver:1.8
        name: echo-app
        ports:
          - containerPort: 8080
            protocol: TCP
        readinessProbe:
          failureThreshold: 10
          httpGet:
            path: /healthz
            port: 8080
            scheme: HTTP
          periodSeconds: 1
          successThreshold: 1
          timeoutSeconds: 1
      # keep `args:` simple for a simpler `kubectl patch` command
      - args:
        - --version
        - $(VERSION)
        command:
        - /usr/sbin/start_esp
        - --http_port
        - "{{.ce_esp_port}}"
        - --backend
        - "127.0.0.1:{{.app_port}}"
        - --service
        - "{{.ce_service_name}}"
        image: gcr.io/endpoints-release/endpoints-runtime:1.9
        name: esp
        ports:
          - containerPort: 8000
            protocol: TCP
```

After `gcloud service-management deploy openapi.yml` and determining the revision of the API, run this command to patch the manifest file in-place.

```
kubectl patch \
  -f .kube.yml \
  --local=true \
  -o yaml \
  -p '{"spec":{"template":{"spec":{"containers":[{"name":"esp","args":["--version","2017-06-28r0"]}]}}}}' \
  > .kube.yml
```

Then continue as normal with the pipeline.

---

> NOTE: `deploy-api` is a program that needs to created based on the changes in this PR (determining the revision to set)

`.drone.yml` (0.5+) would look something like this:

```yml
pipeline:
  cloud-endpoints:
    - image: whatever
    # deploy-api compares and deploys the api spec, printing out the revision
    - export VERSION=$(deploy-api --spec openapi.yml)
    # note: ${VERSION}
    - kubectl patch -f .kube.yml --local=true -o yaml -p '{"spec":{"template":{"spec":{"containers":[{"name":"esp","args":["--version","'${VERSION}'"]}]}}}}'

  drone-gke:
    ...
```

`.drone.yml` (0.4) would look something like this:

```yml
deploy:
  cloud-endpoints:
    image: jmccann/drone-commands
    commands:
      # deploy-api compares and deploys the api spec, printing out the revision
      - export VERSION=$(deploy-api --spec openapi.yml)
      # note: ${VERSION}
      - kubectl patch -f .kube.yml --local=true -o yaml -p '{"spec":{"template":{"spec":{"containers":[{"name":"esp","args":["--version","'${VERSION}'"]}]}}}}'

  drone-gke:
    ...
```

**Comment by jrynyt** [comment]
timing this one out

**Pull request closed without merging**